styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java

```diff
@@ -102,64 +109,67 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl,
+      final String message = String.format("Couldn't remove workflow %s. ", workflowId);
+      LOG.warn(message, e);
+      throw new RuntimeException(message, e);
+    }
+    LOG.info("Workflow removed: {}", workflowId);
+    return Response.forStatus(NO_CONTENT);
   }
 
-  private CompletionStage<Response<ByteString>> createOrUpdateWorkflow(String componentId,
-                                                                       RequestContext rc) {
-    // TODO: handle validation in one place, see SchedulerResource.java
+  private Response<Workflow> createOrUpdateWorkflow(String componentId,
+                                                    RequestContext rc) {
     final Optional<ByteString> payload = rc.request().payload();
     if (!payload.isPresent()) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload.")));
+      return Response.forStatus(BAD_REQUEST.withReasonPhrase("Missing payload."));
     }
     final WorkflowConfiguration workflowConfig;
     try {
       workflowConfig = OBJECT_MAPPER
           .readValue(payload.get().toByteArray(), WorkflowConfiguration.class);
     } catch (IOException e) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST
-                                 .withReasonPhrase("Invalid payload. " + e.getMessage())));
+      return Response.forStatus(BAD_REQUEST
+          .withReasonPhrase("Invalid payload. " + e.getMessage()));
     }
 
-    final Collection<String> errors = workflowValidator.validateWorkflowConfiguration(workflowConfig);
+    final Collection<String> errors =
+        workflowValidator.validateWorkflowConfiguration(workflowConfig);
     if (!errors.isEmpty()) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid workflow configuration: " + errors)));
+      return Response.forStatus(
+          BAD_REQUEST.withReasonPhrase("Invalid workflow configuration: " + errors));
+    }
+
+    final Workflow workflow = Workflow.create(componentId, workflowConfig);
+    final Optional<Workflow> oldWorkflowOptional;
+    try {
+      oldWorkflowOptional = workflowInitializer.inspectChange(workflow);
+    } catch (WorkflowInitializationException e) {
+      return Response.forStatus(BAD_REQUEST.withReasonPhrase(e.getMessage()));
     }
 
-    // TODO: handle workflow crud directly in api service instead of proxying to scheduler
-    return forwardingClient.send(rc.request()
-                                     .withPayload(payload.get())
-                                     .withUri(schedulerApiUrl("workflows", componentId)));
+    if (oldWorkflowOptional.isPresent()) {
+      LOG.info("Workflow modified, old config: {}, new config: {}", oldWorkflowOptional.get(),
+          workflow);
+    } else {
+      LOG.info("Workflow added: {}", workflow);
+    }
+
+    return Response.forPayload(workflow);
```

Comment:
Also call `workflowConsumer.accept(oldWorkflowOptional, Optional.of(workflow));`

Reply:
This needs a bit a rework. This consumer is hook on Scheduler not API, and it requires stats to send metric data, but stats is not available on API. I think we can create an instance of stats for API as well, but not sure what would happen for standalone setup.

Reply:
It is already one service for standalone setup. That’s why I’m not sure having two stats instances. Would that make Apollo sad? Maybe inject a stats factory? For standalone that would return a singleton.

Reply:
Standalone still treats scheduler and api as separate apps. I think we might as well merge them into a single app to make things simpler.

As a smaller step, yes we could just inject the same stats instance into both the scheduler and api.